src/linker/Linker/TypeNameResolver.cs

```diff
@@ -77,6 +77,9 @@ TypeReference ResolveTypeName (AssemblyDefinition assembly, TypeName typeName)
 				if (elementType == null)
 					return null;
 
+				if (typeName is ArrayTypeName)
+					return new ArrayType (elementType);
+
 				return elementType;
```

Comment:
If we were to really fix this it should work for all potential types - so anything which derives from `HasElementType`:
- array
- multi-dim array
- by-ref types
- pointer types